src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs

```diff
@@ -119,6 +121,503 @@ internal int EffectiveMaxConcurrencyLevel
+        }
+
+        public override void Body(TValue value)
+        {
+            _bodyWithStateAndIndex(value, State, State.GetCurrentIteration<TIndex>());
+        }
+    }
+
+    internal abstract class WorkerWithLocal<TValue, TIndex, TLocal> : WorkerBodyWithIndex<TValue, TIndex>
+        where TIndex: INumber<TIndex>
+    {
+        private readonly Action<TLocal>? _localFinally;
+        protected TLocal LocalValue;
+
+        protected WorkerWithLocal(
+            ParallelLoopStateFlags sharedPStateFlags,
+            Func<TLocal> localInit,
+            Action<TLocal>? localFinally) : base(sharedPStateFlags)
+        {
+            LocalValue = localInit();
+            _localFinally = localFinally;
+        }
+
+        public override void Finally()
+        {
+            // If a cleanup function was specified, call it. Otherwise, if the type is
+            // IDisposable, we will invoke Dispose on behalf of the user.
+            _localFinally?.Invoke(LocalValue);
+        }
+    }
+
+    internal sealed class WorkerWithStateAndLocal<TValue, TIndex, TLocal> : WorkerWithLocal<TValue, TIndex, TLocal>
+        where TIndex: INumber<TIndex>
+    {
+        private readonly Func<TValue, ParallelLoopState, TLocal, TLocal> _bodyWithLocal;
+
+        internal WorkerWithStateAndLocal(
+            ParallelLoopStateFlags sharedPStateFlags,
+            Func<TValue, ParallelLoopState, TLocal, TLocal> bodyWithLocal,
+            Func<TLocal> localInit,
+            Action<TLocal>? localFinally) : base(sharedPStateFlags, localInit, localFinally)
+        {
+            _bodyWithLocal = bodyWithLocal;
+        }
+
+        public override void Body(TValue value)
+        {
+            LocalValue = _bodyWithLocal(value, State, LocalValue);
+        }
+    }
+
+    internal sealed class WorkerWithEverything<TValue, TIndex, TLocal> : WorkerWithLocal<TValue, TIndex, TLocal>
+        where TIndex: INumber<TIndex>
+    {
+        private readonly Func<TValue, ParallelLoopState, TIndex, TLocal, TLocal> _bodyWithEverything;
+
+        public WorkerWithEverything(
+            ParallelLoopStateFlags sharedPStateFlags,
+            Func<TValue, ParallelLoopState, TIndex, TLocal, TLocal> bodyWithEverything,
+            Func<TLocal> localInit,
+            Action<TLocal>? localFinally) : base(sharedPStateFlags, localInit, localFinally)
+        {
+            _bodyWithEverything = bodyWithEverything;
+        }
+
+        public override void Body(TValue value)
+        {
+            LocalValue = _bodyWithEverything(value, State, State.GetCurrentIteration<TIndex>(), LocalValue);
+        }
+    }
+
+    #endregion
+
+    #region Worker Bodies Factories
+
+    internal interface IWorkerBodyFactory<in TValue>
+    {
+        IWorkerBody<TValue> CreateWorkerBody(ParallelLoopStateFlags sharedPStateFlags);
+    }
+
+    internal sealed class WorkerWithSimpleBodyFactory<TValue>: IWorkerBodyFactory<TValue>
```

Comment:
I can change all the factories to structs if that is desirable - there is currently no inheritance. 
This means I could change the functions receiving them to be generic over the factory and (theoretically) reduce allocation.

Should I do this?